packages/langium/src/generator/generator-node.ts

```diff
@@ -78,6 +97,64 @@ export class CompositeGeneratorNode {
         return this.contents.length === 0;
     }
 
+    /**
+     * Adds tracing information in form of `{astNode, property?, index?}` to `this` generator node.
+     * Overwrites existing trace data, if set previously.
+     *
+     * The given data are kept as they are, the actual resolution of text positions within the DSL text
+     * is done at the final processing of `this` node as part of {@link toStringAndTrace()}.
+     *
+     * @param astNode the AstNode corresponding to `this` node's content
+     *
+     * @param property the value property name (string) corresponding to `this` node's content,
+     *  e.g. if this node's content corresponds to some `string` or `number` property; is optional
+     *
+     * @param index the index of the value within a list property corresponding to `this` node's content,
+     * if the property contains a list of elements, is ignored otherwise,
+     * must not be given if no `property` is given ; is optinal
+     *
+     * @returns `this` {@link CompositeGeneratorNode} for convenience.
+     */
+    trace<T extends AstNode>(astNode: T, property?: Properties<T>, index?: number): this;
```

Comment:
Minor: We should use multiple overloads for this, so that users cannot enter `property: undefined` with `index` set. We already validate against that anyway, so it would be good to let the overload reflect this.
```suggestion
    trace<T extends AstNode>(astNode: T): this;
    trace<T extends AstNode>(astNode: T, property: Properties<T>, index?: number): this;
```

Reply:
It's not too important. Just wanted to have a quick-type-safety-win if possible. If it leads to other issues, we can keep it as it is for now.

Reply:
It definitely makes sense, I didn't think into this direction. Thanks for this hint!

Reply:
Just pushed an additional commit splitting the previous `(astNode: T, property?: Properties<T>, index?: number)` signatures into pairs of `(astNode: T, property?: Properties<T>)` and `(astNode: T, property: Properties<T>, index: number | undefined)`.

Let's see in action how well e.g. the TS content assist deals with them.

Reply:
Yep, I think that's good :)